[tools/depends] pythonmodule-pil update to 7.1.2

fuzzard opened the pull request:
## Description
updates python Pillow addon to 5.1.0 - inline with Version supplied by Ubuntu Bionic Beaver 18.04 PPA
Have an additional bump to 7.1.2. Both commits are runtime tested individually.

Note:
The rbpi patch is a weird one. Only required for 7.1.2, and its the only platform that doesnt seem to respect either the disable-lcms flag, or the disable-platform-guessing. Both flags are in the logs as expected, but it still tries to compile lcms and fails due to /usr/local usage that should be disabled by the disable-platform-guessing.
I couldnt spot where or how this is happening in setup.py, so patching this out will have to do. Maybe someone else will find time to look into it more in the future.

Also fixes ios/tvos runtime usage

## Motivation and Context
Discussion about some dependency updates on slack

## How Has This Been Tested?
Tested on ios using pil test plugin created by @enen92 https://github.com/enen92/plugin.image.piltest
Had to do the addon.xml update, so the updating test addon can be grabbed from my repo
https://github.com/fuzzard/plugin.image.piltest

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

fuzzard commented:
Will have to upload tarball to mirrors.

@wsnipex not sure if this affects PPA stuff at all? Ive bumped to version supplied by default ubuntu PPA for 18.04. Hope thats suitable

@Rechi i was hoping you could let me know if anything android related needs to be done. I know you have dabbled with a depends update for this (to 7+ from what i saw). Also not entirely sure about the patch, in particular the linux platform part being dropped completely.

wsnipex commented:
Looks sane to me.
Pillow 5.1.0 uploaded to mirrors.

enen92 commented:
Thanks! @sualfred can you do a few tests on Android since your skin/scripts use it?

enen92 commented on addons/script.module.pil/addon.xml:
```diff
@@ -1,7 +1,7 @@
 <addon id="script.module.pil"
        name="Python Image Library"
-       version="1.1.7"
+       version="5.1.0+matrix.1"
```
5.1.0 please, no need to keep the matrix identifier

sualfred commented:
@enen92 
Sure will do it this weekend

Rechi commented:
Why not update to the most recent version instead of using a two year old version?

fuzzard commented:
Honestly, because of PPA stuff. I keep hearing conflicting things around it all. Some say to keep things lowest available on lowest supported ubuntu (18.04 in this case).

From the apple side, we build it all anyways, so doesnt matter to me personally in the slightest, but from what i gather, PPA stuff still uses tools/depends. I guess wsnipex could package up the newer version for the kodi PPA's?

Rechi commented:
PPA doesn't matter for the versions used in depends.
For Kodi code you should keep in mind that there are distributions which do not ship the version from depends. So write code that it is compatible with the currently supported Ubuntu versions. But that doesn't mean depends have to use outdated versions.

Commit:
[tools/depends] bump pythonmodule-pil 7.1.2

Fix ios/tvos usage

rbpi requires to explicitly remove lcms, as it doesnt respect the --disable-lcms
flag for some reason, and causes a build failure trying to build the _imagingcms extension when it shouldnt
be building it at all.

 building 'PIL._imagingcms' extension
 /usr/bin/ccache /home/jenkins/rbpi-dev/tools/arm-bcm2708/arm-rpi-4.9.3-linux-gnueabihf/bin/arm-linux-gnueabihf-gcc -fPIC -mcpu=cortex-a7 -mfloat-abi=hard -mfpu=neon-vfpv4 -mvectorize-with-neon-quad -pipe -mabi=aapcs-linux -Wno-psabi -Wa,-mno-warn-deprecated -Wno-deprecated-declarations -isystem/home/jenkins/rbpi-dev/firmware/opt/vc/include -isystem/home/jenkins/rbpi-dev/firmware/opt/vc/include/interface/vcos/pthreads -isystem/home/jenkins/rbpi-dev/firmware/opt/vc/include/interface/vmcs_host/linux -Og -g -D_DEBUG -isystem /home/jenkins/workspace/LINUX-RBPI/tools/depends/xbmc-depends/raspberry-pi2-debug/include -I/home/jenkins/workspace/LINUX-RBPI/tools/depends/xbmc-depends/raspberry-pi2-debug/include/freetype2 -I/home/jenkins/workspace/LINUX-RBPI/tools/depends/target/pythonmodule-pil/raspberry-pi2-debug/src/libImaging -I/home/jenkins/workspace/LINUX-RBPI/tools/depends/xbmc-depends/i686-linux-gnu-native/include -I/home/jenkins/workspace/LINUX-RBPI/tools/depends/xbmc-depends/raspberry-pi2-debug/include -I/home/jenkins/workspace/LINUX-RBPI/tools/depends/xbmc-depends/raspberry-pi2-debug/include/python3.7 -I/usr/local/include -I/usr/include -I/home/jenkins/workspace/LINUX-RBPI/tools/depends/xbmc-depends/i686-linux-gnu-native/include/python3.7 -c src/_imagingcms.c -o build/temp.linux-x86_64-3.7/src/_imagingcms.o
 In file included from /home/jenkins/workspace/LINUX-RBPI/tools/depends/xbmc-depends/raspberry-pi2-debug/include/python3.7/Python.h:11:0,
                 from src/_imagingcms.c:29:
 /usr/include/limits.h:26:36: fatal error: bits/libc-header-start.h: No such file or directory
   #include <bits/libc-header-start.h>
                                     ^
 compilation terminated.
 error: command '/usr/bin/ccache' failed with exit status 1
 Makefile:71: recipe for target '.installed-raspberry-pi2-debug' failed

fuzzard commented:
jenkins build this please

fuzzard commented:
jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
@Rechi Ive added a commit to bump to 7.1.2. I can/will squash it all, just waiting on a runtime test for an existing issue to be done.

Have updated the patch to a minimal patch required now that Pil has a disable-platform-guessing switch.

Ive cleaned up the crossflags a bit, removing some superfluous stuff. Runtime tests ok, so seems ok.

Theres a bit of duplication in the way both android and darwin_embedded platforms work. I dont know a better way to handle that, so any ideas you may be able to provide would be great.

wsnipex commented:
is this runtime tested on droid? I suspect it's broken once again.

fuzzard commented:
Its not currently, and i assume the breakage will refer to the link flags. Happy to put them back in. Compiled ok without them, but i guess theres something droid specific that im not aware of.

fuzzard commented:
Did a quick runtime using enen's piltest addon against the 5.1.0 jenkins artifact for android x86, and it seemed to run ok. Would love someone to runtime it on an arm based android app.

the 5.1.0 build job is https://jenkins.kodi.tv/job/BuildMulti-PR/16309/ and the android artifacts are based on this without the 7.1.2 bump commit

wsnipex commented:
from a quick look into the apk, this shouldn't work:

```
in /assets/addons/script.module.pil/lib/PIL/_imaging.py
__file__ = pkg_resources.resource_filename(__name__, '_imaging.cpython-37-x86_64-linux-gnu.so')

```
The libraries are renamed and moved during packaging, resulting in 
`/lib/armeabi-v7a/lib_imaging.cpython-37-x86_64-linux-gnu.so`

So, you reverted https://github.com/xbmc/xbmc/pull/17890

fuzzard commented:
No worries, to be honest im not sure how that line was removed completely. Have thrown it back in

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Ok, runtime tested on armv8 android (and the apple platforms). Should be ready now

The rbpi patch is a weird one. Only required for 7.1.2, and its the only platform that doesnt seem to respect either the disable-lcms flag, or the disable-platform-guessing. Both flags are in the logs as expected, but it still tries to compile lcms and fails due to /usr/local usage that should be disabled by the disable-platform-guessing. 
I couldnt spot where or how this is happening in setup.py, so patching this out will have to do. Maybe someone else will find time to look into it more in the future.

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Any further comments/changes?

If not, ill look to merge this tomorrow now ive got green ticks on everything.

Rechi commented on addons/script.module.pil/addon.xml:
```diff
@@ -1,7 +1,7 @@
 <addon id="script.module.pil"
        name="Python Image Library"
-       version="3.0.0"
+       version="7.1.2"
```
At least 7.1.2 is guaranteed for all platforms?

fuzzard replied:
How best to handle this? Nothing's really guaranteed on non depends platforms,.ie anything linux. Ubuntu 18 ships the 5.1.0, do I reduce to that?
Does it really matter what this sits at? Like does it have an affect, or is it just cosmetic in add-on info?

fuzzard replied:
Just dropped back to 5.1.0. Lowest common denominator of supported platforms i guess

Windows - 6.2.1
Apple - 7.1.2
Android - 7.1.2
Linux - whatever they ship i guess. Ubuntu 18.04 - 5.1.0

Rechi commented on tools/depends/target/pythonmodule-pil/01-rbpi-remove-lcms.patch:
```diff
@@ -0,0 +1,31 @@
++#            )
+ 
+         if feature.webp:
+             libs = [feature.webp]
```
new line at the end of the file is missing

Rechi commented on tools/depends/target/pythonmodule-pil/pillow-crosscompile.patch:
```diff
@@ -29,179 +29,4 @@
- 
-         if os.path.isfile("_imagingtiff.c") and feature.tiff:
-             exts.append(Extension(
+         # add platform directories
```
new line at the end of the file is missing

Rechi reviewed (commented):
Having two commits is wrong, as the first one doesn't build on all platforms.

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
Jenkins build this please

Rechi commented on tools/depends/target/pythonmodule-pil/Makefile:
```diff
@@ -41,23 +46,26 @@ $(TARBALLS_LOCATION)/$(ARCHIVE):
 	rm -rf $(PLATFORM)/*; mkdir -p $(PLATFORM)
 	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
+ifeq ($(TARGET_PLATFORM),raspberry-pi)
+	cd $(PLATFORM); patch -p1 -i ../01-rbpi-remove-lcms.patch
```
Why patch the build system instead of using the already existing configure option `--disable-lcms`?

fuzzard replied:
Because it doesnt work on rbpi platform. No idea why. Works fine on every other one.

Rechi commented on tools/depends/target/pythonmodule-pil/01-rbpi-remove-lcms.patch:
```diff
@@ -0,0 +1,31 @@
+--- a/setup.py
++++ b/setup.py
+@@ -752,17 +763,17 @@
```
```suggestion
@@ -752,17 +752,17 @@
```

Rechi commented on tools/depends/target/pythonmodule-pil/Makefile:
```diff
@@ -1,36 +1,41 @@
+CROSSFLAGS=PYTHONXCPREFIX="$(PREFIX)" CC="$(CC) $(CFLAGS)" LDSHARED="$(LDSHARED)" LDFLAGS="$(LDFLAGS) $(EXTRALDFLAGS)" PYTHONPATH="$(PYTHONPATH)"
+
+ifeq ($(CROSS_COMPILING),yes)
+PLATFORM_GUESSING=--disable-platform-guessing
```
This is required for all platforms to get rid of the system include and library paths, which where previously patched away.

fuzzard replied:
I assume what you are saying is all platforms using this dependency are cross_compiling anyway, so just add the flag for everything?
Updated to that assumption, if im wrong, please be more explicit in what you want me to do.

fuzzard commented:
Jenkins build this please
